### PR TITLE
Add used_underscore_binding lint

### PR DESCRIPTION
**Pull request by devonhollowood**
Implement  #488

### CODE REVIEW

**Comment by devonhollowood** [comment]
One thing I noticed is that this lint, as it currently stands, is not quite compatible with the `unused_variables` built-in lint. This can be seen in `tests/compile_fail/for_loop.rs`, line 128:
```
let mut _index = 0;
for _v in &vec { _index += 1 } //~ERROR the variable `_index` is used as a loop counter         
```

Here, we consider the variable `_index` to be used, so we lint on the seconds line. However, the `unused_variables` lint considers `_index` to be unused (because nothing is done with it besides the increment), so if we change `_index` to `index`, it sets off that lint. I'm not sure how best to move our definition of `used` inline with the `unused_variables` lint, but I'm guessing it will involve liveness analysis?

**Comment by Manishearth** [comment]
Looks good so far. I'd like to do something other than checking for `__` though. And we need to see if we can avoid the mismatch with unused_variables. One way of doing that is to use an ExprUseVisitor. Another is to only lint when the path is on the right hand side, i.e. the parent expr is not an ExprAssign or `+=` or whatever with the path on the lhs.

File: src/misc.rs [hunk 1]
```diff
@@ -316,3 +317,68 @@ impl LateLintPass for PatternPass {
...
+
+/// **What it does:** This lint checks for the use of bindings with a single leading underscore
+///
+/// **Why is this bad?** A single leading underscore is usually used to indicate that a binding
+/// will not be used. Using such a binding breaks this expectation.
+///
+/// **Known problems:** This lint's idea of a "used" variable is not quite the same as in the
+/// built-in `unused_variables` lint. For example, in the following code
+/// ```
+/// fn foo(_y: u32) -> u32) {
```
Comment (oli-obk) [inline src/misc.rs]: This should be `y`, not `_y`

File: tests/compile-fail/used_underscore_binding.rs [hunk 2]
```diff
@@ -0,0 +1,30 @@
+#![feature(plugin)]
+#![plugin(clippy)]
+#![deny(clippy)]
+
+fn prefix_underscore(_x: u32) -> u32{
```
Comment (oli-obk) [inline tests/compile-fail/used_underscore_binding.rs]: missing a space between the return type and the opening bracket

File: src/misc.rs [hunk 3]
```diff
@@ -316,3 +317,68 @@ impl LateLintPass for PatternPass {
...
+                let ident = path.segments.last()
+                                .expect("path should always have at least one segment")
+                                .identifier;
+                ident.name.as_str().chars().next() == Some('_') //starts with '_'
+                && ident.name.as_str().chars().skip(1).next() != Some('_') //doesn't start with "__"
+                && ident.name != ident.unhygienic_name //not in macro
+                && cx.tcx.def_map.borrow().values().any(|res| match res.base_def {
+                                                  Def::DefLocal(_, _) => true,
+                                                  _ => false
+                                            }) //local variable
```
Comment (oli-obk) [inline src/misc.rs]: these lines are very oddly aligned and I wonder why the `single_match` lint doesn't warn here about using `if let` instead.

File: src/misc.rs [hunk 4]
```diff
@@ -316,3 +317,68 @@ impl LateLintPass for PatternPass {
...
+    fn check_expr(&mut self, cx: &LateContext, expr: &Expr) {
+        let needs_lint = match expr.node {
+            ExprPath(_, ref path) => {
+                let ident = path.segments.last()
+                                .expect("path should always have at least one segment")
+                                .identifier;
+                ident.name.as_str().chars().next() == Some('_') //starts with '_'
+                && ident.name.as_str().chars().skip(1).next() != Some('_') //doesn't start with "__"
+                && ident.name != ident.unhygienic_name //not in macro
+                && cx.tcx.def_map.borrow().values().any(|res| match res.base_def {
```
Comment (oli-obk) [inline src/misc.rs]: uhm... what's the intended purpose of this statement? If I'm reading this right you are checking whether the `def_map` contains any `DefLocal`s... I think that's true for any non-trivial program...

Comment (Manishearth) [inline src/misc.rs]: Yeah, we should be doing `cx.tcx.def_map.borrow().find(nodeid_of_expr)`

**Comment by devonhollowood** [comment]
Okay, I made @oli-obk's suggested changes. I have a couple questions about @Manishearth's comments:

> I'd like to do something other than checking for __ though.

I think we should still only lint single underscores, but I agree it would be more satisfying to have a better check here. Is there some way to check to see if something appears in a macro definition vs. in its expansion? I think it would make sense to lint the use of underscore-prefixed bindings within a macro definition, but not within its expansion.

> And we need to see if we can avoid the mismatch with unused_variables. One way of doing that is to use an ExprUseVisitor. Another is to only lint when the path is on the right hand side, i.e. the parent expr is not an ExprAssign or += or whatever with the path on the lhs.

I think we probably want to do something more advanced than just checking to make sure the binding never appears on the LHS of an `ExprAssign`. For example, in
```
fn increment(_x: u32) -> u32{
    _x + 1
}
```
we probably want to lint, but `_x` never appears on the LHS of an `ExprAssign`. I'd be happy to try using an `ExprUseVisitor`, but I can't really find any documentation on its use anywhere. I'll try asking around on IRC later.

**Comment by Manishearth** [comment]
> we probably want to lint, but _x never appears on the LHS of an ExprAssign

Exactly! I was saying that we change the heuristics to:

  - We have an ExprPath starting with only one underscore
  - Its parent expr is _not_ an ExprAssign or ExprAssignOp where the path doesn't appear on the RHS

All other Exprs "use" the variable, so we should lint in that case.

The built in analysis is in librustc/middle/liveness.rs (It's a hardwired lint), but IIRC it's intertwined with other analyses, so it's harder to figure out.

If you want to try ExprUseVisitor, you basically create a [Delegate](http://manishearth.github.io/rust-internals-docs/rustc/middle/expr_use_visitor/trait.Delegate.html) (see the escape analysis lint) impl, and check for uses of a variable with the borrow/consume methods. I don't think this is necessary here.

> Is there some way to check to see if something appears in a macro definition vs. in its expansion?

If an Ident's name and hygenic name match then it wasn't created by a macro. I think. This stuff is in flux right now.

**Comment by devonhollowood** [comment]
> Exactly! I was saying that we change the heuristics to:
>
>  * We have an ExprPath starting with only one underscore
>  * Its parent expr is not an ExprAssign or ExprAssignOp where the path doesn't appear on the RHS
>
> All other Exprs "use" the variable, so we should lint in that case

Ah okay that makes sense! I'll go implement that.

> This still doesn't mean it's a local variable; you want to .get(&expr.id) and then match if the obtained def is a deflocal.

I'll go make this change, but I think the fact that I didn't already catch that I'm doing this wrong means my test coverage isn't good enough. I expanded test coverage by adding underscore-prefixed functions, structs, enums, and enum variants. All the above were correctly not linted with the code as is, so could you give me an example of something that the code as-is would lint (due to the `DefLocal` thing), but that it shouldn't lint?

**Comment by Manishearth** [comment]
> something that the code as-is would lint (due to the DefLocal thing), but that it shouldn't lint?

Probably something like `let x = foo_` where `foo_` is a function. The DefLocal check isn't too important, really.

**Comment by devonhollowood** [comment]
> Probably something like let x = foo_ where foo_ is a function.

I added a test for this, and it seems to work with the code as-is. It also seems to work with that check taken out completely.

**Comment by Manishearth** [comment]
Meh, get rid of that check then :)Manishearth/rust-clippy wrote:
Probably something like let x = foo_ where foo_ is a function.


I added a test for this, and it seems to work with the code as-is. It also seems to work with that check taken out completely.

**Comment by devonhollowood** [comment]
Okay, I think the most recent pull requests address all the above issues. If this looks good to you too, I'll go ahead and rebase this all into one commit.

File: src/misc.rs [hunk 5]
```diff
@@ -316,3 +317,69 @@ impl LateLintPass for PatternPass {
...
+        }
+    }
+}
+
+fn is_used(cx: &LateContext, expr: &Expr) -> bool {
+    if let Some(ref parent) = get_parent_expr(cx, expr) {
+        match parent.node {
+            ExprAssign(_, ref rhs) => **rhs == *expr,
+            ExprAssignOp(_, _, ref rhs) => **rhs == *expr,
+            _ => is_used(cx, &parent)
```
Comment (Manishearth) [inline src/misc.rs]: Oh I like this way of calculating is_used; smart :smile: !

File: tests/compile-fail/used_underscore_binding.rs [hunk 6]
```diff
@@ -0,0 +1,81 @@
...
+#![deny(clippy)]
+
+/// Test that we lint if we use a binding with a single leading underscore
+fn prefix_underscore(_foo: u32) -> u32 {
+    _foo + 1 //~ ERROR used binding which is prefixed with an underscore
+}
+
+/// Test that we lint even if the use is within a macro expansion
+fn in_macro(_foo: u32) {
+    println!("{}", _foo); //~ ERROR used binding which is prefixed with an underscore
```
Comment (Manishearth) [inline tests/compile-fail/used_underscore_binding.rs]: \o/

**Comment by Manishearth** [comment]
Great work! Thanks!

**Comment by llogiq** [comment]
I think we may want to reuse some of the techniques in other lints. Great work indeed! :+1:

**Pull request merged**